### DISCUSSION TITLE
Interpreting Raw Yolo Outputs From Tensorrt

### DISCUSSION THREAD

**Discussion by Matt-Tav** [ultralytics/ultralytics#21962] [Q&A] [answered]
## Context

I'm trying to run a yolo11 segmentation model using the tensorrt c++ API. I think I almost have it working, but there's some critical mistake I'm making in decoding the outputs.

### Setup
1. Jetson Xavier NX 8gb
2. Jetpack 5.0.2 L4T 35.1.0 (locked to this jetpack due to sensor drivers)
4. Ultralytics 8.3.61 (most recent version I can install successfully)
5. Tensorrt 8.4.1.5 (most recent version supported by my platform)
6. Python 3.8.10 (most recent version supported by my platform)

### Method
1. Download default weights from ultralytics for `yolo11n-seg.pt`
2. Export model to onnx: `yolo export model=yolo11n-seg.pt format=onnx`
3. Convert onnx to tensorrt engine: `trtexec --onnx=yolo11n-seg.onnx --fp16 --saveEngine=yolo11n-seg-fp16.engine`
4. Verify engine works using `yolo segment predict model=yolo11m-seg-fp16.engine source=bus.jpg data=coco.yaml`
5. Load the same engine and model and attempt inference with my script.

I know the recommended method is to export directly as `.engine`, but the tensorrt API fails to deserialize these direct exports, I think due to the additional metadata inserted by the yolo exporter. As I can get the model to produce correct results using `yolo segment predict` I have been putting off handling the additional meta data for now.

From [polygraphy](https://pypi.org/project/polygraphy/) I understand that the raw inputs/outputs from the model are
- `images  float32 [1,3,640,640]` with `[<batch_size>,(3)channels,h,w]` (row major NCHW?)
- `output0 float32 [1,116,8400]` with `[<batch_size>,<cx,cy,h,w,(n)classes,(32)mask_coeff>,(8400)detections]` (row major?)
- `output1 float32 [1,32,160,160]` with `[<batch_size>,(32)mask_prototypes,h,w]` (row major NCHW?)

Following this understanding I attempt to decode the bounding boxes as follows (you can see the full code and engine file [here](https://github.com/Matt-Tav/tensorrt_experiments))
```c++
    /*
     * Decode detections
     */

    // Output structure [box<4> conf<n_classes> mask_coefs<32>] * 8400
    const int box_offset = 0;
    const int class_offset = 4;                   // xywh
    const int num_classes = 80;                   // coco classes
    const int pred_size = output_det_dims.d[1];   // 116
    const int num_preds = output_det_dims.d[2];   // 8400

    float* det_buffer = output_det_buffer.get();

    // Collect detections based on confidence threshold 
    std::vector<int> det_class;
    std::vector<float> det_conf;
    std::vector<cv::Rect> det_box;
    std::vector<float*> det_ptr;

    float* det_start_ptr; float* conf_start_ptr; float* conf_stop_ptr; float* cls_ptr;
    float confidence_threshold = 0.8;

    for ( int i = 0; i < num_preds; i++ )
    {
        det_start_ptr = det_buffer + (i * pred_size);           // Start of detection
        conf_start_ptr = det_start_ptr + class_offset;          // Start of detection class confidences
        conf_stop_ptr = conf_start_ptr + num_classes;           // End of detection class confidences

        cls_ptr = std::max_element(conf_start_ptr, conf_stop_ptr);      // Pointer to highest confidence class

        if ( *cls_ptr >= confidence_threshold )
        {
            det_class.emplace_back(cls_ptr-conf_start_ptr);     // Store detection class
            det_conf.emplace_back(*cls_ptr);                    // Store detection confidence
            det_ptr.emplace_back(det_start_ptr);                // Store pointer to start of detection

            float cx = *(det_start_ptr + 0);   // center x
            float cy = *(det_start_ptr + 1);   // center y
            float w =  *(det_start_ptr + 2);   // width
            float h =  *(det_start_ptr + 3);   // height
            float x1 = cx - (w / 2.0f);        // top left x 
            float y1 = cy - (h / 2.0f);        // top left y

            cv::circle(letterbox_mat, cv::Point(int(cx*640.0f), int(cy*640.0f)), 2, cv::Scalar(0, 255, 255), -1);

            det_box.emplace_back(            // Store bounding box (de-normalize)
                    int(x1 * 640.0f),        // x
                    int(y1 * 640.0f),        // y
                    int(w  * 640.0f),        // w
                    int(h  * 640.0f)         // h
            );
        }
    }
```

The output is garbage however - You can see just the center points plotted here:
<img width="637" height="641" alt="image" src="https://github.com/user-attachments/assets/b3870f3c-2eac-46fd-9121-45b2ce9bf23c" />

The boxes aren't much better, and non-max suppression doesn't resolve this to a reasonable output. The raw values look reasonable, and stay consistent between runs -  you can see the first few raw values off the buffer here:
```
0.584843
0.645527
0.607124
0.606748
0.58287
0.577538
0.564754
0.571108
0.559359
0.566373
0.558691
0.572759
0.574068
0.579485
```

## Questions
1. Is my basic understanding of the decoding theory correct?
2. Is if there is any kind of post-processing / activation I need to apply?
3. Do I need to take any steps to account for the half precision fp16 conversion?

I've tried to debug myself as much as possible, following #14341 and #19385, but I'm just stuck now:(

Thanks in advance for any help!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Matt-Tav, thank you for your interest in Ultralytics 🚀! This is an automated response to help you get unblocked quickly—an Ultralytics engineer will also assist you here soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it. For your case, this could include:
-Exact export and TensorRT build commands used, plus engine files or logs
-A small input image and the raw output tensors you’re decoding
-The exact code snippet performing post-processing on outputs

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a **Python>=3.8** environment with **PyTorch>=1.8** to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including CUDA/CUDNN, Python and PyTorch preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Reply by glenn-jocher** [reply]
Thanks for the detailed context—your decode logic is close, but the main issue is tensor layout. TensorRT/ONNX exports return `output0` as (N, C=116, S=8400), so you must transpose to (N, S, C) or index with a stride of `S` across channels (Ultralytics’ Python does `pred.transpose(-1, -2)` before postprocess, see the AutoBackend reference). Details: [AutoBackend forward (transpose step)](https://docs.ultralytics.com/reference/nn/autobackend/), and export notes: [TensorRT export for YOLO](https://docs.ultralytics.com/integrations/tensorrt/).

Minimal C++ indexing (no transpose) for detection i:
```cpp
// output0: [1, 116, 8400] -> C=116, S=8400
const int C = 116, S = 8400, NC = 80;
const float* det = output0;  // float32 binding

auto v = [&](int c, int i){ return det[c * S + i]; };  // channel-first

float cx = v(0, i), cy = v(1, i), w = v(2, i), h = v(3, i);  // xywh
int cls = 0; float conf = 0.f;
for (int c = 0; c < NC; ++c) { float p = v(4 + c, i); if (p > conf) { conf = p; cls = c; } }
// convert to xyxy as needed, apply conf/iou thresholds + NMS
```

Notes:
- No extra activations are required; class scores are already post‑sigmoid in exported graphs.
- FP16 engines don’t change decode logic—use the binding data type to allocate buffers; TensorRT will cast as needed.
- For masks, flatten `output1` to (32, 160*160) and compute per‑det `mask = sigmoid(protos @ mask_coeffs)`, then upsample to the letterboxed image and crop by the box.

If issues persist, please share the full text traceback/logs plus `yolo checks` output and a minimal engine + image so we can reproduce.

**Reply by Matt-Tav** [reply]
Hi Glenn, thanks for the quick reply and info!

This [repo](https://github.com/Matt-Tav/tensorrt_experiments) contains the models and code to reproduce my testing, including a [1 for 1 python clone](https://github.com/Matt-Tav/tensorrt_experiments/blob/main/scripts/infer.py) of the code to hopefully make debugging easier.

I've tried substituting your code exactly:
```c++
    /*
     * Decode detections
     */

    const int C = 116, S = 8400, NC = 80;
    const float* det = output_det_buffer.get();  // float32 binding
    auto v = [&](int c, int i){ return det[c * S + i]; };  // channel-first
    
    std::vector<float> det_conf;
    std::vector<cv::Rect> det_box;
    std::vector<int> det_class;
    
    for ( int i = 0; i < S; i++ )
    {
        float cx = v(0, i), cy = v(1, i), w = v(2, i), h = v(3, i);  // xywh
        float x1 = cx - (w / 2.0f);
        float y1 = cy - (h / 2.0f);
        
        int cls = 0; float conf = 0.f;
        for (int c = 0; c < NC; ++c) { float p = v(4 + c, i); if (p > conf) { conf = p; cls = c; } }
        cv::circle(letterbox_mat, cv::Point(int(cx*640.0f), int(cy*640.0f)), 2, cv::Scalar(0, 255, 255), -1);
        
        if ( conf > 0.8 ) {    
            det_class.emplace_back(cls);     // Store class
            det_conf.emplace_back(conf);     // Store confidence
            det_box.emplace_back(            // Store bounding box (de-normalize)
                    int(x1 * 640.0f),        // x
                    int(y1 * 640.0f),        // y
                    int(w  * 640.0f),        // w
                    int(h  * 640.0f)         // h
            );
        }
    }
```
 Unfortunately the output still seems garbled given as it detects in the blank areas.
<img width="320" height="320" alt="image" src="https://github.com/user-attachments/assets/205fb865-b799-4696-90c5-8cb8823de9a1" /> <img width="320" height="320" alt="image" src="https://github.com/user-attachments/assets/dc52e5c8-8361-43a7-87b4-5479655ca1f9" />

The python clone produces a similar result:
<img width="320" height="320" alt="image" src="https://github.com/user-attachments/assets/fe9d10cd-1c0e-4388-bca2-b2a79cf60b70" /> <img width="320" height="320" alt="image" src="https://github.com/user-attachments/assets/45de3aae-b2f7-4502-8a7e-5f19d9804063" />

I'm thinking the issue may actually be in how im preparing the image to pass to the network - I tried copying from `ultralytics/engine/predictor.py`
```python
    flat_image_mat = np.stack([np.asarray(letterbox_mat, dtype=np.float32)])
    flat_image_mat = flat_image_mat[..., ::-1].transpose((0,3,1,2))

    print(flat_image_mat.shape)
    print(flat_image_mat.dtype)
    print(np.amax(flat_image_mat))

    flat_image_mat = np.ascontiguousarray(flat_image_mat)
    flat_image_mat = torch.from_numpy(flat_image_mat)
    flat_image_mat = flat_image_mat.float()
```
which prints exactly what I would expect:
```
(1, 3, 640, 640)
float32
1.0
```
I found by accident however that if I divide the image by 255 twice, the outputs suddenly become almost reasonable? I expect this is just a coincidence though.
<img width="320" height="320" alt="image" src="https://github.com/user-attachments/assets/f7a8cf01-9eb8-459a-a03b-52eb4cd7cf68" /> <img width="320" height="320" alt="image" src="https://github.com/user-attachments/assets/d1f43435-609c-46fb-8ec2-f95b8fef9433" />

The same model and image run with ```yolo segment predict model=yolo11m-seg-fp16.engine source=bus.jpg data=coco.yaml```
<img width="202" height="270" alt="image" src="https://github.com/user-attachments/assets/bf3a54c9-41ce-46fa-8fb7-c1405c22b60d" />

Here is my yolo checks output, let me know if there's anything else I can provide!
```
Ultralytics 8.3.61 🚀 Python-3.8.10 torch-1.13.0a0+d0d6b1f2.nv22.10 CUDA:0 (Xavier, 6855MiB)
Setup complete ✅ (6 CPUs, 6.7 GB RAM, 84.3/916.5 GB disk)

OS                  Linux-5.10.104-tegra-aarch64-with-glibc2.29
Environment         Linux
Python              3.8.10
Install             pip
RAM                 6.69 GB
Disk                84.3/916.5 GB
CPU                 ARMv8 Processor rev 0 (v8l)
CPU count           6
GPU                 Xavier, 6855MiB
GPU count           1
CUDA                11.4

numpy               ✅ 1.23.5>=1.23.0
numpy               ✅ 1.23.5<2.0.0; sys_platform == "darwin"
matplotlib          ✅ 3.7.5>=3.3.0
opencv-python       ✅ 4.10.0.84>=4.6.0
pillow              ✅ 10.4.0>=7.1.2
pyyaml              ✅ 5.3.1>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.10.1>=1.4.1
torch               ✅ 1.13.0a0+d0d6b1f2.nv22.09>=1.8.0
torch               ✅ 1.13.0a0+d0d6b1f2.nv22.09!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.13.0>=0.9.0
tqdm                ✅ 4.66.5>=4.64.0
psutil              ✅ 5.5.1
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.0.3>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.6>=2.0.0
```

**Reply by Matt-Tav** [reply, marked as answer]
Hi Glenn, figured it out!

For some reason the model expects the bindings in the order ```['images', 'output1', 'output0']```
I was passing them as  ```['images', 'output0', 'output1']``` - everything works now, thank you for the help!

<img width="320" height="320" alt="image" src="https://github.com/user-attachments/assets/f8417434-a105-4543-b42f-f7c568c2aa1d" />

**Reply by glenn-jocher** [reply]
Great catch—TensorRT doesn’t guarantee binding order, so always map by binding name and set input dims explicitly for dynamic shapes; a minimal pattern is below, and for reference, `output0` is detections [1,116,8400] and `output1` are mask protos [1,32,160,160] as used in the exporter’s postprocess path documented in the [Exporter reference (forward/postprocess shapes)](https://docs.ultralytics.com/reference/engine/exporter/). 

```cpp
// Resolve by name, not index
int i_images = engine->getBindingIndex("images");
int i_det    = engine->getBindingIndex("output0");
int i_proto  = engine->getBindingIndex("output1");
assert(i_images >= 0 && i_det >= 0 && i_proto >= 0);

// Set input shape for dynamic networks (TRT 8.x)
auto d = engine->getBindingDimensions(i_images);
if (d.d[0] == -1)                      // e.g., [-1,3,640,640]
    context->setBindingDimensions(i_images, nvinfer1::Dims4{1, 3, 640, 640});
```

If you revisit direct `.engine` export, ensure the engine is built with the same TensorRT major/CUDA as your runtime (deser issues are typically version mismatches rather than metadata); build notes live in the [TensorRT export implementation details](https://docs.ultralytics.com/reference/engine/exporter/).
